core/libp2p/muxer/yamux/yamux.hpp

```diff
@@ -0,0 +1,245 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_YAMUX_HPP
+#define KAGOME_YAMUX_HPP
+
+#include <map>
+#include <queue>
+
+#include <boost/asio/streambuf.hpp>
+#include <boost/system/error_code.hpp>
+#include "common/logger.hpp"
+#include "libp2p/common/network_message.hpp"
+#include "libp2p/muxer/yamux/yamux_config.hpp"
+#include "libp2p/stream/stream.hpp"
+#include "libp2p/transport/connection.hpp"
+#include "libp2p/transport/muxed_connection.hpp"
+
+namespace libp2p::stream {
+  class YamuxStream;
+}
+
+namespace libp2p::muxer {
+  class YamuxFrame;
+
+  /**
+   * Implementation of stream multiplexer - connection, which has only one
+   * physical link to another peer, but many logical streams, for example, for
+   * several applications
+   * Read more: https://github.com/hashicorp/yamux/blob/master/spec.md
+   */
+  class Yamux : public transport::MuxedConnection,
+                public std::enable_shared_from_this<Yamux> {
+   public:
+    using StreamId = uint32_t;
+    using NewStreamHandler =
+        std::function<void(std::unique_ptr<stream::Stream>)>;
+
+    enum class YamuxErrorStream {
+      NO_SUCH_STREAM = 1,
+      NOT_WRITABLE,
+      NOT_READABLE,
+      YAMUX_IS_CLOSED
+    };
+
+    /**
+     * Create a new Yamux instance
+     * @param connection, multiplexed by this instance
+     * @param stream_handler - function, which is going to be called, when a new
+     * stream arrives
+     * @param yamux_config to configure this instance
+     */
+    Yamux(
+        std::shared_ptr<transport::Connection> connection,
+        NewStreamHandler stream_handler, YamuxConfig yamux_config,
+        kagome::common::Logger logger = kagome::common::createLogger("Yamux"));
+
+    ~Yamux() override;
+
+    void start() override;
+
+    void stop() override;
+
+    outcome::result<std::unique_ptr<stream::Stream>> newStream() override;
+
+    void close() override;
+
+    bool isClosed() override;
+
+   private:
+    /**
+     * Helper function, which closes the Yamux; needed in order to void
+     * "calling virtual functions from destructor"
+     */
+    void closeYamux();
```

Comment:
Well, yeah, the resources are going to be freed automatically, so I am going to remove this function call from the dtor. But, the function itself is needed in case we want to stop Yamux by ourselves.